physics.control: fixed a bug in rewrite of tf to ss

RaMathuZen opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #26827 
#### Brief description of what is fixed or changed

Refactored the code to fix the issue by scaling  the coefficients of numerator & denominator polynomials when constructing the state space matrices (only for C, D and for A it exists already) and also revised the test in order to check for this bug.

The bug  was because of mismatch in transfer function used in `rewrite` and that of in the docs. (assuming the leading coefficient to be unity)

#### Other comments
Now,
``` python
>>> import sympy
>>> from sympy.physics.control.lti import StateSpace, TransferFunction
>>>
>>> s, K, T1 = sympy.symbols('s, K, T1')
>>>
>>> G = TransferFunction(K, T1*s + 1, s)
>>> print(G)
TransferFunction(K, T1*s + 1, s)
>>>
>>> Gss = G.rewrite(StateSpace)
>>> print(Gss)
StateSpace(Matrix([[-1/T1]]), Matrix([[1]]), Matrix([[K/T1]]), Matrix([[0]]))
>>>
>>> Gss.rewrite(TransferFunction)
[[TransferFunction(K, T1*s + 1, s)]]
```
Previously (which was wrong),
``` python
>>> print(Gss)
StateSpace(Matrix([[-1/T1]]), Matrix([[1]]), Matrix([[K]]), Matrix([[0]]))
```

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- physics.control
  - fixed an issue in incorrect conversion of transfer function to statespace 
<!-- END RELEASE NOTES -->

RaMathuZen committed:
physics.control: fixed a bug in rewrite of tf to ss and revised tests

There was a bug in the rewrite of transfer function(tf) to state space (ss)
model as mentioned in issue #26827, where because of a mismatch in tf model
used in the rewrite implementation and the one in documentation it lead to
an incorrect conversion, i.e. mainly because of the assumption in the tf
model used in the implementation that, the leading coefficent of highest
degree term in denominator polynomial was unity, but in the docs it was
generic. So refactored the code to fix this issue by scaling down the
coefficients when constructing the ss matrices (only for C, D and for A it
exists already) and also revised the test in order to check for this bug.

Example:

>>> import sympy
>>> from sympy.physics.control.lti import StateSpace, TransferFunction
>>>
>>> s, K, T1 = sympy.symbols('s, K, T1')
>>>
>>> G = TransferFunction(K, T1*s + 1, s)
>>> print(G)
TransferFunction(K, T1*s + 1, s)
>>>
>>> Gss = G.rewrite(StateSpace)
>>> print(Gss)
StateSpace(Matrix([[-1/T1]]), Matrix([[1]]), Matrix([[K/T1]]), Matrix([[0]]))
>>>
>>> Gss.rewrite(TransferFunction)
[[TransferFunction(K, T1*s + 1, s)]]

Previously, Gss was (which is wrong),

>>> StateSpace(Matrix([[-1/T1]]), Matrix([[1]]), Matrix([[K]]), Matrix([[0]]))

RaMathuZen committed:
author: add Annamalai N to .mailmap

RaMathuZen committed:
Update .mailmap

RaMathuZen committed:
Update .mailmap - 2

RaMathuZen committed:
Update .mailmap-3

RaMathuZen committed:
Update lti.py

moorepants commented:
You'll need to address the code quality failure so the rest of the tests run.

RaMathuZen committed:
refactored code for passing code quality, pypy3.10 test checks

---

physics.control: fixed a bug in rewrite of tf to ss and revised tests

    There was a bug in the rewrite of transfer function(tf) to state space (ss)
    model as mentioned in issue #26827, where because of a mismatch in tf model
    used in the rewrite implementation and the one in documentation it lead to
    an incorrect conversion, i.e. mainly because of the assumption in the tf
    model used in the implementation that, the leading coefficent of highest
    degree term in denominator polynomial was unity, but in the docs it was
    generic. So refactored the code to fix this issue by scaling down the
    coefficients when constructing the ss matrices (only for C, D and for A it
    exists already) and also revised the test in order to check for this bug.

    Example:

    >>> import sympy
    >>> from sympy.physics.control.lti import StateSpace, TransferFunction
    >>>
    >>> s, K, T1 = sympy.symbols('s, K, T1')
    >>>
    >>> G = TransferFunction(K, T1*s + 1, s)
    >>> print(G)
    TransferFunction(K, T1*s + 1, s)
    >>>
    >>> Gss = G.rewrite(StateSpace)
    >>> print(Gss)
    StateSpace(Matrix([[-1/T1]]), Matrix([[1]]), Matrix([[K/T1]]), Matrix([[0]]))
    >>>
    >>> Gss.rewrite(TransferFunction)
    [[TransferFunction(K, T1*s + 1, s)]]

    Previously, Gss was (which is wrong),

    >>> StateSpace(Matrix([[-1/T1]]), Matrix([[1]]), Matrix([[K]]), Matrix([[0]]))

RaMathuZen commented:
> You'll need to address the code quality failure so the rest of the tests run.

Is it fine now ?

moorepants commented:
I just tested this example:
![image](https://github.com/user-attachments/assets/124b7511-66ca-4985-b7b9-8b090408241b)

and I get:


```python
In [1]: import sympy as sm

In [2]: from sympy.physics.control.lti import StateSpace, TransferFunction

In [3]: w, z = sm.symbols('omega, zeta', real=True)

In [4]: s = sm.symbols('s')

In [5]: t = TransferFunction(w**2, s**2 + 2*z*w*s + w**2, s)

In [6]: t
Out[6]: TransferFunction(omega**2, omega**2 + 2*omega*s*zeta + s**2, s)

In [7]: ss = t.rewrite(StateSpace)

In [8]: ss
Out[8]: 
StateSpace(
Matrix([
[        0,             1],
[-omega**2, -2*omega*zeta]]),

Matrix([
[0],
[1]]),

Matrix([[omega**2, 0]]),

Matrix([[0]]))
```

That did not give the B and C matrices I would expect.

moorepants commented:
What "form" is the rewrite applying? controller form, observer form, etc?

RaMathuZen commented:
> What "form" is the rewrite applying? controller form, observer form, etc?

To Controllable Canonical Form and that's what the function previously does. The example you provided is in observable form.

moorepants commented:
You need a test that shows both C and D are working. There is no test for D.

RaMathuZen committed:
updated the test & fixed the code

---

physics.control: fixed a bug in rewrite of tf to ss and revised tests

        There was a bug in the rewrite of transfer function(tf) to state space (ss)
        model as mentioned in issue #26827, where because of a mismatch in tf model
        used in the rewrite implementation and the one in documentation it lead to
        an incorrect conversion, i.e. mainly because of the assumption in the tf
        model used in the implementation that, the leading coefficent of highest
        degree term in denominator polynomial was unity, but in the docs it was
        generic. So refactored the code to fix this issue by scaling down the
        coefficients when constructing the ss matrices (only for C, D and for A it
        exists already) and also revised the test in order to check for this bug.

        Example:

        >>> import sympy
        >>> from sympy.physics.control.lti import StateSpace, TransferFunction
        >>>
        >>> s, k, tau, p  = sympy.symbols('s, k, tau, p')
        >>>
        >>> G = TransferFunction(k*s + p, tau*s + 1, s)
        >>> G
        TransferFunction(k*s + p, s*tau + 1, s)
        >>>
        >>> G.rewrite(StateSpace)
        StateSpace(
        Matrix([[-1/tau]]),

        Matrix([[1]]),

        Matrix([[-k/tau**2 + p/tau]]),

        Matrix([[k/tau]]))
        >>> Gss = G.rewrite(StateSpace)
        >>>
        >>> Gss.rewrite(TransferFunction)
        [[TransferFunction(k*s + p, s*tau + 1, s)]]
        >>>

RaMathuZen commented:
> You need a test that shows both C and D are working. There is no test for D.

A nice catch. Thanks. Have updated it now. Any other ?  Is the observable canonical form conversion needed ?

moorepants commented:
I just tried https://cpjobling.github.io/eglm03-textbook/07/4/canon.html#system-with-a-proper-transfer-function with your PR and get:

```python
In [1]: import sympy as sm

In [2]: from sympy.physics.control import TransferFunction, StateSpace

In [3]: c0, c1, c2 = sm.symbols('c0, c1, c2', real=True)

In [4]: a0, a1, a2, a3 = sm.symbols('a0, a1, a2, a3', real=True)

In [5]: d = sm.symbols('d', real=True)

In [6]: s = sm.symbols('s')

In [11]: t = TransferFunction(d, 1, s) + TransferFunction(c2*s**2 + c1*s**1 + c0, s**4 + a3*s**3 + a2*s**2 + a1*s**1 + a0, s)

In [13]: sm.pprint(t)
                          2       
d         c₀ + c₁⋅s + c₂⋅s        
─ + ──────────────────────────────
1                   2       3    4
    a₀ + a₁⋅s + a₂⋅s  + a₃⋅s  + s 

In [14]: t.rewrite(StateSpace)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[14], line 1
----> 1 t.rewrite(StateSpace)

File ~/src/sympy/sympy/core/basic.py:2091, in Basic.rewrite(self, deep, *args, **hints)
   2088         return self
   2089 # hereafter, empty pattern is interpreted as all pattern.
-> 2091 return self._rewrite(pattern, rule, method, **hints)

File ~/src/sympy/sympy/core/basic.py:2096, in Basic._rewrite(self, pattern, rule, method, **hints)
   2094 deep = hints.pop('deep', True)
   2095 if deep:
-> 2096     args = [a._rewrite(pattern, rule, method, **hints)
   2097             for a in self.args]
   2098 else:
   2099     args = self.args

File ~/src/sympy/sympy/core/basic.py:2096, in <listcomp>(.0)
   2094 deep = hints.pop('deep', True)
   2095 if deep:
-> 2096     args = [a._rewrite(pattern, rule, method, **hints)
   2097             for a in self.args]
   2098 else:
   2099     args = self.args

File ~/src/sympy/sympy/core/basic.py:2103, in Basic._rewrite(self, pattern, rule, method, **hints)
   2101 meth = getattr(self, method, None)
   2102 if meth is not None:
-> 2103     rewritten = meth(*args, **hints)
   2104 else:
   2105     rewritten = self._eval_rewrite(rule, args, **hints)

File ~/src/sympy/sympy/physics/control/lti.py:940, in TransferFunction._eval_rewrite_as_StateSpace(self, *args)
    938 a = den_coeffs[1:]
    939 a_mat = Matrix([[(-1)*coefficient/den_coeffs[0] for coefficient in reversed(a)]])
--> 940 vert = zeros(n-1, 1)
    941 mat = eye(n-1)
    942 A = vert.row_join(mat)

File ~/src/sympy/sympy/matrices/dense.py:1093, in zeros(*args, **kwargs)
   1090 if 'c' in kwargs:
   1091     kwargs['cols'] = kwargs.pop('c')
-> 1093 return Matrix.zeros(*args, **kwargs)

File ~/src/sympy/sympy/matrices/matrixbase.py:1205, in MatrixBase.zeros(kls, rows, cols, **kwargs)
   1203     cols = rows
   1204 if rows < 0 or cols < 0:
-> 1205     raise ValueError("Cannot create a {} x {} matrix. "
   1206                      "Both dimensions must be positive".format(rows, cols))
   1207 klass = kwargs.get('cls', kls)
   1208 rows, cols = as_int(rows), as_int(cols)

ValueError: Cannot create a -1 x 1 matrix. Both dimensions must be positive
```

moorepants commented:
This works though:

```python
In [19]: t.doit().rewrite(StateSpace)
Out[19]: 
StateSpace(
Matrix([
[  0,   1,   0,   0],
[  0,   0,   1,   0],
[  0,   0,   0,   1],
[-a0, -a1, -a2, -a3]]),

Matrix([
[0],
[0],
[0],
[1]]),

Matrix([[c0, c1, c2, 0]]),

Matrix([[d]]))
```

RaMathuZen committed:
updated the lti.py and its test for few other edge cases
---
physics.control: fixed a bug in rewrite of tf to ss and revised tests

        There was a bug in the rewrite of transfer function(tf) to state space (ss)
        model as mentioned in issue #26827, where because of a mismatch in tf model
        used in the rewrite implementation and the one in documentation it lead to
        an incorrect conversion, i.e. mainly because of the assumption in the tf
        model used in the implementation that, the leading coefficent of highest
        degree term in denominator polynomial was unity, but in the docs it was
        generic. So refactored the code to fix this issue by scaling down the
        coefficients when constructing the ss matrices (only for C, D and for A it
        exists already) and also revised the test in order to check for this bug.

        Example:

        >>> import sympy
        >>> from sympy.physics.control.lti import StateSpace, TransferFunction
        >>>
        >>> s, k, tau, p  = sympy.symbols('s, k, tau, p')
        >>>
        >>> G = TransferFunction(k*s + p, tau*s + 1, s)
        >>> G
        TransferFunction(k*s + p, s*tau + 1, s)
        >>>
        >>> G.rewrite(StateSpace)
        StateSpace(
        Matrix([[-1/tau]]),

        Matrix([[1]]),

        Matrix([[-k/tau**2 + p/tau]]),

        Matrix([[k/tau]]))
        >>> Gss = G.rewrite(StateSpace)
        >>>
        >>> Gss.rewrite(TransferFunction)
        [[TransferFunction(k*s + p, s*tau + 1, s)]]
        >>>

RaMathuZen commented:
> I just tried https://cpjobling.github.io/eglm03-textbook/07/4/canon.html#system-with-a-proper-transfer-function with your PR and get:

Now I've updated it so that it can handle these zero-order transfer functions and also the unity transfer function ones.

moorepants commented:
I haven't had time to look at it further. I'm vacation now.

RaMathuZen commented:
> I haven't had time to look at it further. I'm vacation now.

Ok

moorepants commented:
```
In [1]: %paste
In [1]: import sympy as sm

In [2]: from sympy.physics.control import TransferFunction, StateSpace

In [3]: c0, c1, c2 = sm.symbols('c0, c1, c2', real=True)

In [4]: a0, a1, a2, a3 = sm.symbols('a0, a1, a2, a3', real=True)

In [5]: d = sm.symbols('d', real=True)

In [6]: s = sm.symbols('s')

In [11]: t = TransferFunction(d, 1, s) + TransferFunction(c2*s**2 + c1*s**1 + c0, s**4 + a3*s**3 + a2*s**2 + a1*s**1 + a0, s)

In [13]: sm.pprint(t)

## -- End pasted text --
                          2       
d         c₀ + c₁⋅s + c₂⋅s        
─ + ──────────────────────────────
1                   2       3    4
    a₀ + a₁⋅s + a₂⋅s  + a₃⋅s  + s 

In [2]: t.rewrite(StateSpace)
Out[2]: 
Parallel(StateSpace(
Matrix([[0]]),

Matrix([[0]]),

Matrix([[0]]),

Matrix([[d]])), StateSpace(
Matrix([
[  0,   1,   0,   0],
[  0,   0,   1,   0],
[  0,   0,   0,   1],
[-a0, -a1, -a2, -a3]]),

Matrix([
[0],
[0],
[0],
[1]]),

Matrix([[c0, c1, c2, 0]]),

Matrix([[0]])))
```

Works now.

moorepants commented:
There are now merge conflicts. If you resolve them, I will merge this.

moorepants committed:
Fixed merge conflict.

moorepants committed:
Add test case for Parallel.

moorepants commented on sympy/physics/control/tests/test_lti.py:
```diff
@@ -38,6 +38,34 @@
+                [0],
+                [1]]),
+        Matrix([[0, c0, c1, c2, 0]]),
+        Matrix([[d]]))
```
Is this the solution we expect? It has a zero row in A and B. And zero column in C.

RaMathuZen replied:
@moorepants I found it like If we do rewrite after doit (i.e. `t.doit().rewrite(StateSpace)`) then it is correct but the vice versa gives this wrong solution(i.e. `t.rewrite(StateSpace).doit()`). So it has to do something with doit().

moorepants committed:
Add XFAIL to the Parallel.doit() issue.
